README

      ARAnyM (Atari Running on Any Machine)
      version 1.0.2 released on 2014/10/17


 1) License
 ----------

    ARAnyM is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    ARAnyM is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with ARAnyM; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA



 2) What is this?
 ----------------

ARAnyM is a multiplatform virtual machine (a software layer, or an emulator)
for running Atari ST/TT/Falcon operating systems and applications on almost
any hardware with many host operating systems.
The reason for writing ARAnyM is to provide Atari power users with
faster and better machines. The ultimate goal is to create a new platform
where TOS/GEM applications could continue to live forever.

 Features:
 ---------
   o 68040 CPU (including MMU040)
   o 68040 and 68881/2 FPU
   o 14 MB ST-RAM and up to 3824 MB (configurable) of FastRAM
   o VIDEL, Blitter, MFP, SCC, ACIA, IKBD for highest possible compatibility
   o Sound (compatible with Atari Falcon DMA Sound)
   o Atari floppy DD/HD for connecting floppy image or real floppy drive
   o two IDE channels for connecting disk images, harddrives or CD-ROMs
   o extended keyboard and mouse support (including mouse wheel)
   o direct access to host file system via BetaDOS and MiNT xfs drivers
   o networking using ethernet emulation with a driver for MiNT-Net
   o TOS 4.04, EmuTOS, MagiC or Linux-m68k as the boot operating system
   o runs with FreeMiNT, MagiC, and any other operating system that runs
     also on real Atari computer.
   o Native CD-ROM access (under Linux, other OS: audio CD only), without
     scsi/ide/whatsoever emulation
   o USB and PCI NatFeat support for developing Atari drivers on ARAnyM

Important: ARAnyM is not finished. Consider this to be a beta version
and a work-in-progress. Sorry for incomplete and confusing documentation.
Any help with programming/documentation/porting is very welcome!


 3) Installing
 -------------

See INSTALL file and https://github.com/aranym/aranym/wiki


 4) Configuring and running
 --------------------------

See https://github.com/aranym/aranym/wiki


 5) Keyboard shortcuts
 ---------------------

Keys not found on a PC keyboard:
--------------------------------

Atari Help key - mapped on the F11
Atari Undo key - mapped on the F12

Keys not found on Atari keyboard:
---------------------------------

Page Up key    - mapped as Shift+Arrow Up (usual combination in GEM apps)
Page Down key  - mapped as Shift+Arrow Down (usual combo in GEM apps)

Special keys (most can be re-defined in the SETUP):
-------------
Pause/Break    - invoke SETUP GUI (ARAnyM on-the-fly configuration)
Shift+Pause    - QUIT ARAnyM
Ctrl+Pause     - REBOOT ARAnyM
Alt+Pause      - enter integrated debugger (only if started with "-D")

PrintScreen    - create a screenshot in actual directory

Alt+Ctrl+Shift+Esc - release the keyboard/mouse input focus so you can use
                 your keyboard and mouse in other host applications.
                 Can be changed to other key combo in the Settings.
                 Middle mouse does the same job.

Scroll Lock    - switch between windowed and fullscreen mode


 10) Troubleshooting
 ------------------

If your Microsoft IntelliMouse doesn't work correctly on Linux framebuffer
console you may want to set the SDL_MOUSEDEV_IMPS2 environment variable to 1.

Example for bash: export SDL_MOUSEDEV_IMPS2=1; aranym


 11) More information
 --------------------

 Read the NEWS file for user visible changes.

 Read ChangeLog for internal changes.
 Look at TODO if you want to help us.

 Join our mailing list for ARAnyM users - the WEB interface for the list
 is at https://lists.bobek.cz/mailman/listinfo/cz-bobek-lists-aranym-user

 Visit https://aranym.github.io/ for latest information and 
 https://github.com/aranym/aranym for source code.

 Help us improve the Wiki documentation at https://github.com/aranym/aranym/wiki

 For developers we have a developer mailing list and a CVS server.
 For details go to https://aranym.github.io/


The "Power Without The Price" is back!


NEWS
 New features:
 -------------
 * action to be taken on middle mouse button press can now be configured
 * new "pipe" setting for output written to the parallel port,
   allowing to pass it eg. directly to lpr
 * NatFeat configuration support for EmuTOS


2019/04/14 - version 1.1.0 released
==========

 List of changes in this release is 60 pages long! No wonder as it's been
 four and half years since previous release, and the development never
 stopped, mainly thanks to Thorsten. See the 'ChangeLog' file for details.

 Tons of things have been fixed in all areas of the ARAnyM, and lots of
 new features have been added. Following list is incomplete but still gives
 you a good overview:

 Notable changes:
 ----------------
 * ARAnyM is hosted by GitHub.com at https://aranym.github.io/
 * there are automated builds for x86/amd64, armhf and macOS
 * added hotkey for turning audio on/off
 * HostFS MiNT file locking implemented
 * nfosmesa updated to OpenGL 4.6
 * MetaDOS driver improved
 * hotkeys work in the Setup GUI dialog as well
 * configurable directory for screenshots (via Setup GUI)
 * on macOS the configuration file is now looked up first
   in ~/Documents/Aranym_files (if it does not exist yet)
 * displaying non-ASCII characters in the fileselector
   should work now
 * path of loaded TOS/EmuTOS can now be selected in the Setup GUI
 * JIT CPU compiler now supports Solaris
 * JIT CPU compiler now supports more 64bit hosts (like kfreebsd)

 New features:
 -------------
 * NatFeat SCSI driver for Linux
 * NatFeat host exec
 * NatFeat for query/set CPU JIT
 * NatFeat for ARAnyM exit
 * new command line option --option
   that can set any value from the config file
 * AppImage can now be built


2014/10/17 - version 1.0.2 released
==========

 o typo error hopefully fixed


2014/10/17 - version 1.0.1 released
==========

 Just a quick bugfix release:

 o SDL NF CDROM is included in the build correctly
 o JIT is supported on kfreebsd (alternate kernel for Debian OS)
 o some warnings that prevented build on Debian fixed


2014/10/15 - version 1.0.0 released
==========

 Major highlights of this release:

 o JIT CPU compiler now supports also ARM platforms driven by Linux

 o JIT CPU compiler fixes memory management and adds some instructions

 o massive NatFeat HostFS fixes and improvements (symlinks and more)

 o NatFeat CD-ROM for Win32 and SDL

 o NatFeat OSMesa support improved (OpenGL 4.4 / Mesa 10.1.4)
   osmesa.ldg can replace tinygl.ldg now

 o seamless mouse integration and keyboard input improved

 o NVRAM: many country codes added (from current FreeMiNT)

 o libSDL2 is supported now (along with libSDL 1.2.x)

 o Cygwin, MinGW, Mac OS X, CentOS and Debian build fixes

Countless bug fixes and many other under-the-hood improvements are included
as well. This release crowns more than 13 years of continuous development
by many dedicated programmers (listed in the AUTHORS file). Thank you all!


2014/04/12 - version 0.9.16 released @ atariada.cz
==========

 Major highlights of this release:

 o JIT CPU compiler supported on 64-bit Linux and Mac OS X now!
   (Jens made a miracle)

 o MFPR FPU emulation should be perfect
   (Andreas ironed out few remainings bugs)

 o Ethernet support under Mac OS X should be way better
   (Philipp added support for big packets and multi-packets)

 o new config setting "LoadToFastRAM" to load kernel in FastRAM
   (is Andreas working on fixing Linux-m68k issues in FastRAM?)

 o David Gálvez improved NatFeat USB support (now requires FreeMiNT 1.19+)

 There's also a small set of bugs fixed in this release - mainly in Mac OS X
 thanks to Philipp but also some generic things like GPIP (Patrice).


2013/04/15 - version 0.9.15 released
==========

 The main new feature of this release is the seamless mouse integration mode:

 o Mouse input handling is improved a lot. Now ARAnyM has a seamless mouse
   integration when new Clocky (2013/04/15) is loaded.

 o There is no need to manually ungrab or grab the mouse.

 o Also mouse grab on left button click now reacts on the first button
   click everytime (previously only if you had window focus) and the
   host and mouse cursor positions are synced.

 o And if you ungrab the mouse intentionally then it's not automatically
   grabbed when the host mouse pointer returns to ARAnyM window.

 o At last, the annoying "resetting weird mouse packet" error is gone, too.

 Bugs fixed:

 o Native Features were missing proper memory overflow checks and bugs
   in NatFeat drivers could kill the ARAnyM.

 o NFVDI had a bug (a race condition) in resolution switching.

 o build date of EmuTOS was decoded improperly


2012/09/20 - version 0.9.14 released
==========

 Mac users: check out the new bridged ethernet (see README_ethernet.rtf).

 All users: try out the new Single Blit Composing and Single Blit Refresh
 modes - they might give you a better video performance.

 There are also many compilation and build fixes for various MacOSX targets
 and few MPFR FPU fixes.

 And last but not least, there's a neat user requested feature - mouse ungrab
 (frees your keyboard and mouse from ARAnyM window so you can work in other
 applications while ARAnyM is running) simply by clicking the host middle
 mouse button (well, works only if you don't have the Eiffel mouse wheel
 emulation enabled in settings).

 
2012/03/23 - version 0.9.13 released
==========

 A couple of important bugfixes accumulated in last few days:

   o USB NatFeat doesn't fail to claim devices in certain cases

   o MPFR FPU got fix for restarting of instructions after access error

   o JIT CPU version should work even when compiled with newer compilers

   o update your mmusetup.cnf in MiNT setup for larger FastRAM (see
     provided atari/mmusetup.cnf that is set for 512 MB FastRAM now)


2012/03/19 - version 0.9.12 released
==========

 New features:

   o MPFR based new precise FPU emulation core. Somewhat slower than the

src/md5.cpp
/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 *
 * Changed so as no longer to depend on Colin Plumb's `usual.h' header
 * definitions; now uses stuff from dpkg's config.h.
 *  - Ian Jackson <ijackson@nyx.cs.du.edu>.
 * Still in the public domain.
 * 
 * Converted to C++ and added all public functions (*buf and FILE*)
 *  - Petr Stehlik <pstehlik@sophics.cz>.
 * Still in the public domain.
 */

#include "sysdeps.h"
#include "aramd5.h"
#include "SDL_compat.h"

# include <cstring>

using namespace std;

#if SDL_BYTEORDER == SDL_BIG_ENDIAN
void
byteSwap(UWORD32 *buf, unsigned words)
{
	md5byte *p = (md5byte *)buf;

	do {
		*buf++ = (UWORD32)((unsigned)p[3] << 8 | p[2]) << 16 |
			((unsigned)p[1] << 8 | p[0]);
		p += 4;
	} while (--words);
}
#else
#define byteSwap(buf,words)
#endif /* SDL_BYTEORDER == SDL_BIG_ENDIAN */

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */

void MD5::MD5Init()
{
	MD5Context *ctx = this->context;

	ctx->buf[0] = 0x67452301;
	ctx->buf[1] = 0xefcdab89;
	ctx->buf[2] = 0x98badcfe;
	ctx->buf[3] = 0x10325476;

	ctx->bytes[0] = 0;
	ctx->bytes[1] = 0;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void
MD5::MD5Update(md5byte const *buf, unsigned len)
{
	UWORD32 t;

	MD5Context *ctx = this->context;

	/* Update byte count */

	t = ctx->bytes[0];
	if ((ctx->bytes[0] = t + len) < t)
		ctx->bytes[1]++;	/* Carry from low to high */

	t = 64 - (t & 0x3f);	/* Space available in ctx->in (at least 1) */
	if (t > len) {
		memcpy((md5byte *)ctx->in + 64 - t, buf, len);
		return;
	}
	/* First chunk is an odd size */
	memcpy((md5byte *)ctx->in + 64 - t, buf, t);
	byteSwap(ctx->in, 16);
	MD5Transform(ctx->buf, ctx->in);
	buf += t;
	len -= t;

	/* Process data in 64-byte chunks */
	while (len >= 64) {
		memcpy(ctx->in, buf, 64);
		byteSwap(ctx->in, 16);
		MD5Transform(ctx->buf, ctx->in);
		buf += 64;
		len -= 64;
	}

	/* Handle any remaining bytes of data. */
	memcpy(ctx->in, buf, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void
MD5::MD5Final(md5byte digest[16])
{
	MD5Context *ctx = this->context;

	int count = ctx->bytes[0] & 0x3f;	/* Number of bytes in ctx->in */
	md5byte *p = (md5byte *)ctx->in + count;

	/* Set the first char of padding to 0x80.  There is always room. */
	*p++ = 0x80;

	/* Bytes of padding needed to make 56 bytes (-8..55) */
	count = 56 - 1 - count;

	if (count < 0) {	/* Padding forces an extra block */
		memset(p, 0, count + 8);
		byteSwap(ctx->in, 16);
		MD5Transform(ctx->buf, ctx->in);
		p = (md5byte *)ctx->in;
		count = 56;
	}
	memset(p, 0, count);
	byteSwap(ctx->in, 14);

	/* Append length in bits and transform */
	ctx->in[14] = ctx->bytes[0] << 3;
	ctx->in[15] = ctx->bytes[1] << 3 | ctx->bytes[0] >> 29;
	MD5Transform(ctx->buf, ctx->in);

	byteSwap(ctx->buf, 4);
	memcpy(digest, ctx->buf, 16);
	memset(ctx, 0, sizeof(MD5Context));	/* In case it's sensitive */
}

#ifndef ASM_MD5

/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#define MD5STEP(f,w,x,y,z,in,s) \
	 (w += f(x,y,z) + in, w = (w<<s | w>>(32-s)) + x)

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 longwords of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
void
MD5::MD5Transform(UWORD32 buf[4], UWORD32 const in[16])
{
	register UWORD32 a, b, c, d;

	a = buf[0];
	b = buf[1];
	c = buf[2];
	d = buf[3];

	MD5STEP(F1, a, b, c, d, in[0] + 0xd76aa478, 7);
	MD5STEP(F1, d, a, b, c, in[1] + 0xe8c7b756, 12);
	MD5STEP(F1, c, d, a, b, in[2] + 0x242070db, 17);
	MD5STEP(F1, b, c, d, a, in[3] + 0xc1bdceee, 22);
	MD5STEP(F1, a, b, c, d, in[4] + 0xf57c0faf, 7);
	MD5STEP(F1, d, a, b, c, in[5] + 0x4787c62a, 12);
	MD5STEP(F1, c, d, a, b, in[6] + 0xa8304613, 17);
	MD5STEP(F1, b, c, d, a, in[7] + 0xfd469501, 22);
	MD5STEP(F1, a, b, c, d, in[8] + 0x698098d8, 7);
	MD5STEP(F1, d, a, b, c, in[9] + 0x8b44f7af, 12);
	MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1, 17);
	MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7be, 22);
	MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122, 7);
	MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193, 12);
	MD5STEP(F1, c, d, a, b, in[14] + 0xa679438e, 17);
	MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821, 22);

	MD5STEP(F2, a, b, c, d, in[1] + 0xf61e2562, 5);
	MD5STEP(F2, d, a, b, c, in[6] + 0xc040b340, 9);
	MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51, 14);
	MD5STEP(F2, b, c, d, a, in[0] + 0xe9b6c7aa, 20);
	MD5STEP(F2, a, b, c, d, in[5] + 0xd62f105d, 5);
	MD5STEP(F2, d, a, b, c, in[10] + 0x02441453, 9);
	MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681, 14);
	MD5STEP(F2, b, c, d, a, in[4] + 0xe7d3fbc8, 20);
	MD5STEP(F2, a, b, c, d, in[9] + 0x21e1cde6, 5);
	MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6, 9);
	MD5STEP(F2, c, d, a, b, in[3] + 0xf4d50d87, 14);
	MD5STEP(F2, b, c, d, a, in[8] + 0x455a14ed, 20);
	MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905, 5);
	MD5STEP(F2, d, a, b, c, in[2] + 0xfcefa3f8, 9);
	MD5STEP(F2, c, d, a, b, in[7] + 0x676f02d9, 14);
	MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8a, 20);

	MD5STEP(F3, a, b, c, d, in[5] + 0xfffa3942, 4);
	MD5STEP(F3, d, a, b, c, in[8] + 0x8771f681, 11);
	MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122, 16);
	MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380c, 23);
	MD5STEP(F3, a, b, c, d, in[1] + 0xa4beea44, 4);
	MD5STEP(F3, d, a, b, c, in[4] + 0x4bdecfa9, 11);
	MD5STEP(F3, c, d, a, b, in[7] + 0xf6bb4b60, 16);
	MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70, 23);
	MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6, 4);
	MD5STEP(F3, d, a, b, c, in[0] + 0xeaa127fa, 11);
	MD5STEP(F3, c, d, a, b, in[3] + 0xd4ef3085, 16);
	MD5STEP(F3, b, c, d, a, in[6] + 0x04881d05, 23);
	MD5STEP(F3, a, b, c, d, in[9] + 0xd9d4d039, 4);
	MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5, 11);
	MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8, 16);
	MD5STEP(F3, b, c, d, a, in[2] + 0xc4ac5665, 23);

	MD5STEP(F4, a, b, c, d, in[0] + 0xf4292244, 6);
	MD5STEP(F4, d, a, b, c, in[7] + 0x432aff97, 10);
	MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7, 15);
	MD5STEP(F4, b, c, d, a, in[5] + 0xfc93a039, 21);
	MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3, 6);
	MD5STEP(F4, d, a, b, c, in[3] + 0x8f0ccc92, 10);
	MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47d, 15);
	MD5STEP(F4, b, c, d, a, in[1] + 0x85845dd1, 21);
	MD5STEP(F4, a, b, c, d, in[8] + 0x6fa87e4f, 6);
	MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0, 10);
	MD5STEP(F4, c, d, a, b, in[6] + 0xa3014314, 15);
	MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1, 21);
	MD5STEP(F4, a, b, c, d, in[4] + 0xf7537e82, 6);
	MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235, 10);
	MD5STEP(F4, c, d, a, b, in[2] + 0x2ad7d2bb, 15);
	MD5STEP(F4, b, c, d, a, in[9] + 0xeb86d391, 21);

	buf[0] += a;
	buf[1] += b;
	buf[2] += c;
	buf[3] += d;
}

void MD5::computeSum(md5byte const *buf, unsigned long size, unsigned char digest[16])
{
	MD5Init();
	MD5Update(buf, size);
	MD5Final(digest);
}

bool MD5::compareSum(md5byte const *buf, unsigned long size, const unsigned char digest[16])
{
	unsigned char buf_digest[16];
	computeSum(buf, size, buf_digest);
	if (memcmp(digest, buf_digest, 16) != 0)
		return false;

	return true;
}

bool MD5::computeSum(FILE *f, unsigned char digest[16])
{
	unsigned char buf[1024];
	int n;

	MD5Init();
	rewind(f);
	while ((n = fread(buf, 1, sizeof(buf), f)) > 0)
		MD5Update(buf, n);
	MD5Final(digest);
	if (ferror(f))
		return false;
	rewind(f);
	return true;
}

bool MD5::compareSum(FILE *f, const unsigned char digest[16])
{
	unsigned char file_digest[16];
	if (! computeSum(f, file_digest))
		return false;

	if (memcmp(digest, file_digest, 16) != 0)
		return false;

	return true;
}

void MD5::printSum(const unsigned char *p)
{
	int i;

	for (i = 0; i < 16; ++i)
		printf("%02x", *p++);
}

bool MD5::printSum(FILE *f)
{
	unsigned char file_digest[16];

	if (! computeSum(f, file_digest))
		return false;

	printSum(file_digest);
	return true;
}

#endif
